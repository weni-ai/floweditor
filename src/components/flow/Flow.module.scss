@import 'variables.module.scss';

.new_updates {

  > div {
    padding: $unnnic-spacing-stack-lg $unnnic-spacing-stack-md !important;
  }

  .content {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 20px;

    .title {
      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-black;
      line-height: $unnnic-line-height-md + $unnnic-font-size-title-sm;
    }

    .description {
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }


    .news_list {
      display: flex;
      gap: $unnnic-spacing-xs;
      flex-direction: column;
      align-items: center;
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }

    .footer {
      color: $unnnic-color-neutral-black;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }
  }

  .buttons {
    margin-top: $unnnic-spacing-stack-lg;
    display: flex;
    gap: $unnnic-spacing-stack-md;
    flex: 1;

    > * {
      flex: 1;

      button {
        width: 100%;
      }
    }
    
    .primary {
      background-color: $unnnic-color-brand-weni-soft;
    }

  }
}