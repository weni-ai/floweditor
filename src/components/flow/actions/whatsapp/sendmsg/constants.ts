import i18n from 'config/i18n';
import { UnnnicSelectOption } from 'components/form/select/SelectElement';

export const MAX_LIST_ITEMS_COUNT = 10;
export const MAX_REPLIES_COUNT = 3;

export const enum WhatsAppMessageType {
  SIMPLE = 'simple',
  INTERACTIVE = 'interactive',
}

export const enum WhatsAppHeaderType {
  MEDIA = 'media',
  TEXT = 'text',
}

export const enum WhatsAppInteractionType {
  LIST = 'list',
  REPLIES = 'replies',
  LOCATION = 'location',
  CTA = 'cta_url',
  FLOW = 'flow_msg',
}

export const WHATSAPP_MESSAGE_TYPE_SIMPLE: UnnnicSelectOption<
  WhatsAppMessageType
> = {
  label: i18n.t('whatsapp_messages.simple_message', 'Simple message'),
  value: WhatsAppMessageType.SIMPLE,
};

export const WHATSAPP_MESSAGE_TYPE_INTERACTIVE: UnnnicSelectOption<
  WhatsAppMessageType
> = {
  label: i18n.t('whatsapp_messages.interactive_message', 'Interactive message'),
  value: WhatsAppMessageType.INTERACTIVE,
};

export const WHATSAPP_HEADER_TYPE_TEXT: UnnnicSelectOption<
  WhatsAppHeaderType
> = {
  label: i18n.t('whatsapp_headers.text', 'Text'),
  value: WhatsAppHeaderType.TEXT,
};
export const WHATSAPP_HEADER_TYPE_MEDIA: UnnnicSelectOption<
  WhatsAppHeaderType
> = {
  label: i18n.t('whatsapp_headers.media', 'Media'),
  value: WhatsAppHeaderType.MEDIA,
};

export const WHATSAPP_INTERACTION_TYPE_LIST: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: WhatsAppInteractionType.LIST,
  label: i18n.t('whatsapp_interactions.list', 'Option list'),
  description: i18n.t(
    'whatsapp_interactions.list_description',
    'Make choices easier with a list of up to 10 interactive options',
  ),
};

export const WHATSAPP_INTERACTION_TYPE_NONE: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: null,
  label: i18n.t('whatsapp_interactions.none', 'No interactions selected'),
};

export const WHATSAPP_INTERACTION_TYPE_REPLIES: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: WhatsAppInteractionType.REPLIES,
  label: i18n.t('whatsapp_interactions.replies', 'Quick replies'),
  description: i18n.t(
    'whatsapp_interactions.replies_description',
    'Create up to 3 quick replies with predefined messages',
  ),
};

export const WHATSAPP_INTERACTION_TYPE_CTA: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: WhatsAppInteractionType.CTA,
  label: i18n.t('whatsapp_interactions.add_url', 'URL Button'),
  description: i18n.t(
    'whatsapp_interactions.add_url_description',
    'Add a button with a link to an external URL.',
  ),
};

export const WHATSAPP_INTERACTION_TYPE_WHATSAPP_FLOWS: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: WhatsAppInteractionType.FLOW,
  label: i18n.t('whatsapp_flows.title', 'WhatsApp Flows'),
  description: i18n.t(
    'whatsapp_flows.description',
    'Use your interactive forms created on Meta.',
  ),
};

export const WHATSAPP_INTERACTION_TYPE_LOCATION: UnnnicSelectOption<
  WhatsAppInteractionType
> = {
  value: WhatsAppInteractionType.LOCATION,
  label: i18n.t('whatsapp_interactions.request_location', 'Request Location'),
  description: i18n.t(
    'whatsapp_interactions.location_description',
    'Ask users for their location to facilitate service',
  ),
};

export const WHATSAPP_MESSAGE_TYPE_OPTIONS: UnnnicSelectOption<
  WhatsAppMessageType
>[] = [WHATSAPP_MESSAGE_TYPE_SIMPLE, WHATSAPP_MESSAGE_TYPE_INTERACTIVE];

export const WHATSAPP_HEADER_TYPE_OPTIONS: UnnnicSelectOption<
  WhatsAppHeaderType
>[] = [WHATSAPP_HEADER_TYPE_MEDIA, WHATSAPP_HEADER_TYPE_TEXT];

export const WHATSAPP_INTERACTION_TYPE_OPTIONS: UnnnicSelectOption<
  WhatsAppInteractionType
>[] = [
  WHATSAPP_INTERACTION_TYPE_NONE,
  WHATSAPP_INTERACTION_TYPE_REPLIES,
  WHATSAPP_INTERACTION_TYPE_LIST,
  WHATSAPP_INTERACTION_TYPE_LOCATION,
  WHATSAPP_INTERACTION_TYPE_CTA,
  WHATSAPP_INTERACTION_TYPE_WHATSAPP_FLOWS,
];
