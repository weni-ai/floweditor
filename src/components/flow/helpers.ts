import { Type, Types } from 'config/interfaces';
import { getType, getTypeConfig } from 'config/typeConfigs';
import { NodeEditorSettings } from 'store/nodeEditor';
import { Asset } from 'store/flowContext';
import { FlowIssue } from 'flowTypes';

export const determineTypeConfig = (nodeSettings: NodeEditorSettings): Type => {
  const renderNode = nodeSettings.originalNode;
  const node = renderNode && renderNode.node;

  if (nodeSettings.originalAction && nodeSettings.originalAction.type) {
    return getTypeConfig(nodeSettings.originalAction.type);
  } else if (node && node.actions && node.actions.length > 0) {
    return getTypeConfig(node.actions[node.actions.length - 1].type);
  } else {
    try {
      const type = getType(renderNode);
      const config = getTypeConfig(type);
      if (config.type !== Types.missing) {
        return config;
      }
      // tslint:disable-next-line:no-empty
    } catch (Error) {}
  }

  // Account for ghost nodes
  if (node && node.router) {
    return getTypeConfig(node.router.type);
  }

  throw new Error(`Couldn't determine type config for: ${node.uuid}`);
};

export const hasIssues = (
  issues: FlowIssue[],
  translating: boolean,
  language: Asset,
): boolean => {
  if (!issues || issues.length === 0) {
    return false;
  }

  if (!translating) {
    return !!issues.find((issue: FlowIssue) => !issue.language);
  } else {
    return !!issues.find((issue: FlowIssue) => issue.language === language.id);
  }
};
