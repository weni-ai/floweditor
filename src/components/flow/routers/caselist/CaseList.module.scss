@import 'src/variables.module.scss';

.case_list_container {
  overflow: hidden;
  border-radius: $curvature;
  margin-top: $unnnic-spacing-stack-xs + $unnnic-spacing-stack-nano;

  &:focus {
    outline: none;
  }

  &.scrolling {
    margin-top: 12px;
    background: rgb(250, 250, 250);
  }

  .case_list {
    display: flex;
    flex-direction: column;
    position: relative;
    overflow-y: auto;
    max-height: 190px;
    row-gap: $unnnic-spacing-stack-xs + $unnnic-spacing-stack-nano;
  }
}

.card_z_index {
  z-index: 2000;
}

.kase {
  &:last-child {
    :global {
      .fe-x {
        opacity: 0 !important;
        cursor: default;
      }
    }

    .dnd_icon {
      opacity: 0;
    }
  }

}