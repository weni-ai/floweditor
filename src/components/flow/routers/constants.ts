import i18n from 'config/i18n';

export const WAIT_LABEL = i18n.t(
  'forms.message_label',
  'If the message response...',
);
export const EXPRESSION_LABEL = i18n.t(
  'forms.expression_label',
  'If the expression...',
);
export const GROUP_LABEL = i18n.t(
  'forms.split_by_groups',
  "Select the groups you'd like to split by below",
);

export const OPERAND_LOCALIZATION_DESC = i18n.t(
  'forms.localize_rules',
  'Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.',
);

// We intentionally don't localize these names
export enum DefaultExitNames {
  All_Responses = 'All Responses',
  No_Response = 'No Response',
  Any_Value = 'Any Value',
  Other = 'Other',
}
