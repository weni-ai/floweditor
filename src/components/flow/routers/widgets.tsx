import * as React from 'react';
import OptionalTextInput from 'components/form/optionaltext/OptionalTextInput';
import { StringEntry } from 'store/nodeEditor';
import i18n from 'config/i18n';

export const createResultNameInput = (
  value: StringEntry,
  onChange: (value: string) => void,
): JSX.Element => {
  return (
    <OptionalTextInput
      name={i18n.t('forms.save_result_name')}
      maxLength={64}
      value={value}
      onChange={onChange}
      toggleText={i18n.t('forms.save_as_title', 'Save as')}
    />
  );
};
