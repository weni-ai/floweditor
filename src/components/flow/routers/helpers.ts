import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { DefaultExitNames } from 'components/flow/routers/constants';
import { Operators, Types } from 'config/interfaces';
import {
  Action,
  CallResthook,
  CallWebhook,
  CallExternalService,
  Case,
  Category,
  Exit,
  FlowNode,
  Router,
  RouterTypes,
  SwitchRouter,
  TransferAirtime,
  UIConfig,
  WebhookExitNames,
  CallClassifier,
  OpenTicket,
  CallWeniGPT,
  SendWhatsAppProduct
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { createUUID, snakify } from 'utils';

export interface CategorizedCases {
  cases: Case[];
  categories: Category[];
  exits: Exit[];
  caseConfig: { [uuid: string]: any };
}
export interface ResolvedRoutes extends CategorizedCases {
  defaultCategory: string;
  timeoutCategory?: string;
}

export const createRenderNode = (
  uuid: string,
  router: Router,
  exits: Exit[],
  type: Types,
  actions: Action[] = [],
  uiConfig: { [key: string]: any } = {}
): RenderNode => {
  const renderNode: RenderNode = {
    node: {
      uuid,
      actions,
      router,
      exits
    },
    ui: {
      type,
      position: null,
      config: uiConfig
    },
    inboundConnections: {}
  };
  return renderNode;
};

export const hasCases = (node: FlowNode): boolean => {
  if (
    node.router &&
    (node.router as SwitchRouter).cases &&
    (node.router as SwitchRouter).cases.length
  ) {
    return true;
  }
  return false;
};

export const getCategories = (renderNode: RenderNode): Category[] => {
  if (renderNode.node.router) {
    return renderNode.node.router.categories;
  }
  return [];
};

export const createCaseProps = (cases: Case[], renderNode: RenderNode): CaseProps[] => {
  const categories: Category[] = getCategories(renderNode);

  // make sure we don't mutate the in memory version of our cases
  const newCases = JSON.parse(JSON.stringify(cases));
  return newCases.map((kase: Case) => {
    const matchingCategory = categories.find(
      (category: Category) => category.uuid === kase.category_uuid
    );

    if (isRelativeDate(kase.type)) {
      if (renderNode.ui.config && renderNode.ui.config.cases) {
        const caseConfig = renderNode.ui.config.cases[kase.uuid];
        if (caseConfig && caseConfig.arguments) {
          kase.arguments = caseConfig.arguments;
        }
      }
    }

    return {
      uuid: kase.uuid,
      kase,
      categoryName: matchingCategory ? matchingCategory.name : null,
      valid: true
    };
  });
};

export const isRelativeDate = (operatorType: Operators): boolean => {
  return !![Operators.has_date_eq, Operators.has_date_gt, Operators.has_date_lt].find(
    (type: string) => operatorType === type
  );
};

const isCategoryMatch = (cat: Category, kase: CaseProps) => {
  // see if we have the same name
  return cat.name.trim() === kase.categoryName.trim();
};

/**
 * Given a set of cases and previous categories, determines correct merging of cases
 * and the union of categories
 */
export const categorizeCases = (
  newCases: CaseProps[],
  originalNode: FlowNode
): CategorizedCases => {
  const categories: Category[] = [];
  const cases: Case[] = [];
  const exits: Exit[] = [];
  const caseConfig: UIConfig = {};

  const originalRouter = originalNode && originalNode.router;
  const previousCategories = (originalRouter && originalRouter.categories) || [];

  // look over the new cases and match up categories and exits
  for (const newCase of newCases) {
    // ignore empty cases
    if (!newCase.categoryName || newCase.categoryName.trim().length === 0) {
      continue;
    }

    // convert relative dates to expressions with configs
    if (isRelativeDate(newCase.kase.type)) {
      caseConfig[newCase.uuid] = { arguments: newCase.kase.arguments };
      newCase.kase.arguments = [`@(datetime_add(today(), ${newCase.kase.arguments[0]}, "D"))`];
    }

    //  see if it exists on a previous case
    let category = categories.find((cat: Category) => isCategoryMatch(cat, newCase));

    // if not, see if that category exists on our old node
    if (!category) {
      category = previousCategories.find((cat: Category) => isCategoryMatch(cat, newCase));

      // still no category, lets see if we can find a case uuid match
      if (!category) {
        const router = getSmartOrSwitchRouter(originalNode);
        if (router) {
          const previousCase = router.cases.find((kase: Case) => kase.uuid === newCase.uuid);
          if (previousCase) {
            const previousCategory = previousCategories.find(
              (cat: Category) => cat.uuid === previousCase.category_uuid
            );

            if (previousCategory) {
              category = { ...previousCategory, name: newCase.categoryName };
            }
          }
        }
      }

      // don't pull over an old category more than once
      if (category && categories.find((cat: Category) => cat.uuid === category.uuid)) {
        category = null;
      }

      // we found an old category, bring it and its exit over
      if (category) {
        categories.push(category);
        const previousExit = originalNode.exits.find(
          (exit: Exit) => category.exit_uuid === exit.uuid
        );
        exits.push(previousExit);
      }
    }

    // if still no category, finally lets just create a new one
    if (!category) {
      const exit: Exit = {
        uuid: createUUID()
      };

      exits.push(exit);

      category = {
        uuid: createUUID(),
        name: newCase.categoryName,
        exit_uuid: exit.uuid
      };

      categories.push(category);
    }

    // lastly, add our case
    cases.push({
      ...newCase.kase,
      category_uuid: category.uuid
    });
  }

  return { cases, categories, exits, caseConfig };
};

export const getSmartOrSwitchRouter = (node: FlowNode): SwitchRouter => {
  if (
    node &&
    node.router &&
    (node.router.type === RouterTypes.switch || node.router.type === RouterTypes.smart)
  ) {
    return node.router as SwitchRouter;
  }
  return null;
};

/**
 * Adds a default route, reusing the previous one if possible
 * @param originalNode
 */
export const getDefaultRoute = (
  defaultCategoryName: string,
  originalNode: FlowNode
): { defaultCategory: Category; defaultExit: Exit } => {
  const originalRouter = getSmartOrSwitchRouter(originalNode);

  // use the previous default if it had one
  if (originalRouter) {
    const defaultCategory = originalRouter.categories.find(
      (cat: Category) => cat.uuid === originalRouter.default_category_uuid
    );

    const defaultExit = originalNode.exits.find((e: Exit) => e.uuid === defaultCategory.exit_uuid);

    defaultCategory.name = defaultCategoryName;

    return { defaultCategory, defaultExit };
  }
  // otherwise, create a new exit and category
  else {
    const defaultExit: Exit = {
      uuid: createUUID()
    };

    const defaultCategory = {
      uuid: createUUID(),
      name: defaultCategoryName,
      exit_uuid: defaultExit.uuid
    };

    return { defaultCategory, defaultExit };
  }
};

const getTimeoutRoute = (
  originalNode: FlowNode
): { timeoutCategory: Category; timeoutExit: Exit } => {
  let timeoutCategory: Category = null;
  let timeoutExit: Exit = null;

  const originalRouter = getSmartOrSwitchRouter(originalNode);

  // see if our previous node had a timeout case
  if (originalRouter) {
    if (originalRouter.wait && originalRouter.wait.timeout) {
      const previousCategory = originalRouter.wait.timeout.category_uuid;
      timeoutCategory = originalRouter.categories.find(
        (cat: Category) => cat.uuid === previousCategory
      );
      timeoutExit = originalNode.exits.find(
        (exit: Exit) => exit.uuid === timeoutCategory.exit_uuid
      );
    }
  }

  if (!timeoutCategory) {
    // create a new route
    timeoutExit = {
      uuid: createUUID()
    };

    timeoutCategory = {
      uuid: createUUID(),
      name: DefaultExitNames.No_Response,
      exit_uuid: timeoutExit.uuid
    };
  }

  return { timeoutCategory, timeoutExit };
};

/**
 * Given a set of cases and previous categories, determines correct merging of cases
 * and the union of categories
 */
export const resolveRoutes = (
  newCases: CaseProps[],
  hasTimeout: boolean,
  originalNode: FlowNode,
  defaultCategoryName: string = null
): ResolvedRoutes => {
  const resolved = categorizeCases(newCases, originalNode);

  let resolvedDefaultCategory = defaultCategoryName;
  if (!resolvedDefaultCategory) {
    resolvedDefaultCategory =
      resolved.categories.length > 0 ? DefaultExitNames.Other : DefaultExitNames.All_Responses;
  }

  // tack on our other category
  const { defaultCategory, defaultExit } = getDefaultRoute(resolvedDefaultCategory, originalNode);
  resolved.categories.push(defaultCategory);
  resolved.exits.push(defaultExit);

  const results: ResolvedRoutes = {
    ...resolved,
    defaultCategory: defaultCategory.uuid
  };

  // add in a timeout route if we need one
  if (hasTimeout) {
    const { timeoutCategory, timeoutExit } = getTimeoutRoute(originalNode);
    resolved.categories.push(timeoutCategory);
    resolved.exits.push(timeoutExit);
    results.timeoutCategory = timeoutCategory.uuid;
  }

  return results;
};

export const createWebhookBasedNode = (
  action:
    | CallWebhook
    | CallResthook
    | OpenTicket
    | TransferAirtime
    | CallExternalService
    | CallWeniGPT
    | SendWhatsAppProduct,
  originalNode: RenderNode,
  useCategoryTest: boolean
): RenderNode => {
  const exits: Exit[] = [];
  let cases: Case[] = [];
  let categories: Category[] = [];

  // see if we are editing an existing router so we reuse exits
  if (
    originalNode &&
    originalNode.node.actions.length === 1 &&
    originalNode.node.actions[0].type === action.type
  ) {
    const previousRouter = getSmartOrSwitchRouter(originalNode.node);
    originalNode.node.exits.forEach((exit: any) => exits.push(exit));
    previousRouter.cases.forEach(kase => cases.push(kase));
    originalNode.node.router.categories.forEach(category => categories.push(category));
  } else {
    // Otherwise, let's create some new ones
    exits.push(
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      }
    );

    categories = [
      {
        uuid: createUUID(),
        name: WebhookExitNames.Success,
        exit_uuid: exits[0].uuid
      },
      {
        uuid: createUUID(),
        name: WebhookExitNames.Failure,
        exit_uuid: exits[1].uuid
      }
    ];

    cases = [
      {
        uuid: createUUID(),
        type: useCategoryTest ? Operators.has_category : Operators.has_only_text,
        arguments: [WebhookExitNames.Success],
        category_uuid: categories[0].uuid
      }
    ];
  }

  let operand = '@results.' + snakify(action.result_name);
  if (!useCategoryTest) {
    operand += '.category';
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    operand: operand,
    cases,
    categories,
    default_category_uuid: categories[categories.length - 1].uuid
  };

  let splitType = Types.split_by_webhook;
  if (action.type === Types.call_resthook) {
    splitType = Types.split_by_resthook;
  } else if (action.type === Types.open_ticket) {
    splitType = Types.split_by_ticket;
  } else if (action.type === Types.transfer_airtime) {
    splitType = Types.split_by_airtime;
  } else if (action.type === Types.call_external_service) {
    splitType = Types.split_by_external_service;
  } else if (action.type === Types.call_wenigpt) {
    splitType = Types.split_by_wenigpt;
  } else if (action.type === Types.send_msg_catalog) {
    splitType = Types.split_by_whatsapp_product;
  }

  return createRenderNode(originalNode.node.uuid, router, exits, splitType, [action]);
};

export const createSplitOnActionResultNode = (
  action: CallClassifier,
  originalNode: RenderNode
): RenderNode => {
  const splitType = Types.split_by_intent;
  const exits: Exit[] = [];
  const router: SwitchRouter = {
    cases: [],
    operand: '',
    categories: [],
    type: RouterTypes.switch,
    default_category_uuid: null
  };

  return createRenderNode(originalNode.node.uuid, router, exits, splitType, [action]);
};
