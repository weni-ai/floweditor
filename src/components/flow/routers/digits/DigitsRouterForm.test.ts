import { RouterFormProps } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { HintTypes, RouterTypes, SwitchRouter, WaitTypes } from 'flowTypes';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createRenderNode, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { shallowToJson } from 'enzyme-to-json';

import DigitsRouterForm from './DigitsRouterForm';
import { createUUID } from 'utils';
import { DefaultExitNames } from 'components/flow/routers/constants';

const exits = [{ uuid: utils.createUUID() }];
const categories = [
  {
    name: DefaultExitNames.All_Responses,
    uuid: utils.createUUID(),
    exit_uuid: exits[0].uuid,
  },
];
const { setup } = composeComponentTestUtils<RouterFormProps>(
  DigitsRouterForm,
  getRouterFormProps(
    createRenderNode({
      actions: [],
      exits,
      router: {
        default_category_uuid: categories[0].uuid,
        type: RouterTypes.switch,
        categories,
        wait: {
          type: WaitTypes.msg,
          hint: { type: HintTypes.digits },
        },
        cases: [],
      },
      ui: { position: { left: 0, top: 0 }, type: Types.wait_for_response },
    }),
  ),
);

mock(utils, 'createUUID', utils.seededUUIDs());

describe(DigitsRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('initializes', () => {
    const { wrapper } = setup(true, {
      nodeSettings: {
        $set: {
          originalNode: createRenderNode({
            actions: [],

            exits: [{ destination_uuid: null, uuid: 'generated_uuid_1' }],
            router: {
              type: RouterTypes.switch,
              operand: DEFAULT_OPERAND,
              categories: [
                {
                  uuid: createUUID(),
                  name: 'Other',
                  exit_uuid: 'generated_uuid_1',
                },
              ],
              cases: [
                {
                  uuid: 'generated_uuid_2',
                  type: Operators.has_any_word,
                  arguments: ['red'],
                  category_uuid: null,
                },
              ],
              default_category_uuid: 'generated_uuid_1',
              result_name: 'Color',
            } as SwitchRouter,
            ui: {
              position: { left: 0, top: 0 },
              type: Types.split_by_expression,
            },
          }),
        },
      },
    });

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: vi.fn(), onClose: vi.fn() },
      });

      instance.handleUpdateResultName('Favorite Color');
      instance.handleCasesUpdated([
        {
          kase: { type: Operators.has_any_word, arguments: ['red'] },
          categoryName: 'Red',
        },
        {
          kase: { type: Operators.has_any_word, arguments: ['maroon'] },
          categoryName: 'Red',
        },
        {
          kase: { type: Operators.has_any_word, arguments: ['green'] },
          categoryName: 'Green',
        },
      ] as CaseProps[]);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: vi.fn(), onClose: vi.fn() },
      });

      instance.handleUpdateResultName('Dont save me bro!');
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });
});
