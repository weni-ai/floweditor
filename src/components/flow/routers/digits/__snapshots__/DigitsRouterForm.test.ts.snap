// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DigitsRouterForm > initializes 1`] = `ShallowWrapper {}`;

exports[`DigitsRouterForm > should render 1`] = `ShallowWrapper {}`;

exports[`DigitsRouterForm > updates > should save changes 1`] = `
{
  "cases": [
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "red",
        ],
        "type": "has_any_word",
      },
    },
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "maroon",
        ],
        "type": "has_any_word",
      },
    },
    {
      "categoryName": "Green",
      "kase": {
        "arguments": [
          "green",
        ],
        "type": "has_any_word",
      },
    },
  ],
  "resultName": {
    "validationFailures": [],
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`DigitsRouterForm > updates > should save changes 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [],
          "exits": [
            {
              "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            },
            {
              "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            },
            {
              "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "red",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "maroon",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "green",
                ],
                "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                "type": "has_any_word",
              },
            ],
            "categories": [
              {
                "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "name": "Red",
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
              {
                "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "name": "Green",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
              {
                "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
                "name": "Other",
                "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
              },
            ],
            "default_category_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "operand": "@input.text",
            "result_name": "Favorite Color",
            "type": "switch",
            "wait": {
              "hint": {
                "type": "digits",
              },
              "type": "msg",
            },
          },
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
        "ui": {
          "config": {
            "cases": {},
          },
          "position": null,
          "type": "wait_for_response",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
